onnx/inliner/inliner.cc

```diff
@@ -246,103 +200,114 @@ class Specializer {
+  // override the VisitGraph method.
+  void VisitGraph(GraphProto* graph) override {
     rename_scopes.emplace_back();
-    for (auto& x : *graph.mutable_input())
+    for (auto& x : *graph->mutable_input())
       Rename(*x.mutable_name());
-    for (auto& init : *graph.mutable_initializer())
+    for (auto& init : *graph->mutable_initializer())
       Rename(*init.mutable_name());
-    for (auto& y : *graph.mutable_output())
+    for (auto& y : *graph->mutable_output())
       Rename(*y.mutable_name());
-    for (auto& n : *graph.mutable_node())
-      Transform(n);
+    for (auto& n : *graph->mutable_node())
+      VisitNode(&n);
     rename_scopes.pop_back();
   }
 
  public:
-  // The main specialization method: specialize a FunctionProto for a particular call-site.
+  // Renames variables in a FunctionProto for inlining a particular call-site. This does the following:
+  // (i)  Rename all intermediate variables in the function to ensure that they are unique (wrt the main graph).
+  // (ii) Rename inputs and outputs using names of actual parameters.
   static void
-  Specialize(const NodeProto& callnode, FunctionProto& callee, std::string unique_suffix, NameGenerator& generator) {
-    AttributeMap map;
-    for (auto& attr : callnode.attribute()) {
-      map[attr.name()] = &attr;
-    }
-    auto lookup = [&](const std::string& name) -> const AttributeProto* {
-      auto iter = map.find(name);
-      return (iter != map.end()) ? iter->second : nullptr;
-    };
-    Specializer specializer(unique_suffix, generator, lookup);
+  Rename(const NodeProto& callnode, FunctionProto& callee, std::string unique_suffix, NameGenerator& generator) {
+    InliningRenamer renamer(std::move(unique_suffix), generator);
 
-    specializer.Bind<false>(*callee.mutable_input(), callnode.input());
-    specializer.Bind<true>(*callee.mutable_output(), callnode.output());
+    renamer.Bind<false>(*callee.mutable_input(), callnode.input());
+    renamer.Bind<true>(*callee.mutable_output(), callnode.output());
 
-    for (auto& n : *callee.mutable_node())
-      specializer.Transform(n);
+    renamer.VisitFunction(&callee);
   }
 };
 
 // Identify the set of all "input" variables used by a given node.
 // This includes the variables listed as node.input, as well as
 // implicit inputs referred to in any graph-valued-attribute of the node.
+// In the case of variables referenced in sub-graphs, only non-local variables
+// are treated as implicit inputs.
 
-struct ComputeUsedVars : public NodeVisitorBase {
-  std::vector<std::string> result;
+class ComputeInputs : private Visitor {
+ private:
+  std::vector<std::unordered_set<std::string>> namescopes;
+
+  bool InNestedScope() const {
+    return !namescopes.empty();
+  }
+
+  std::unordered_set<std::string>& CurrentScope() {
+    return namescopes.back();
+  }
+
+  bool IsLocalVar(const std::string& name) const {
+    for (auto& scope : namescopes) {
+      if (scope.count(name) > 0)
+        return true;
+    }
```

Comment:
```suggestion
      if (scope.count(name) > 0) {
        return true;
      }
    }
```

code robustness against editing errors. @jcwchen I wonder if we should update the clangformat config?

Reply:
Personally I vote for always using bracket, but I don't have strong preference. Feel free to update the config and see if anyone has any concern.

Reply:
Done (for this instance; not to config)